expr/js_eval.go

```diff
@@ -0,0 +1,211 @@
+package expr
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/robertkrimen/otto"
+)
+
+// EvaluateJavascript can evaluate a source Javascript program having set the given
+// subject into the `$` variable.
+func EvaluateJavascript(ctx context.Context, source string, subject any) (result otto.Value, err error) {
+	var halted bool
+	defer func() {
+		if caught := recover(); caught != nil {
+			if halted {
+				err = errors.Wrap(errors.New("timed out executing Javascript code"), err.Error())
+			} else {
+				panic(caught) // it wasn't our interrupt handler, repanic
+			}
+		}
+	}()
+
+	// Create a Javascript virtual machine that we'll use for evaluating the source
+	// expression. We must be very careful: this is executing code on behalf of others, so
+	// comes with all normal warnings.
+	vm := otto.New()
+	vm.Interrupt = make(chan func(), 1)
+
+	// Start a new function bounded context.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// If we haven't finished execution after our timeout, we trigger the interrupt handler.
+	SafelyGo(func() {
+		select {
+		case <-time.After(250 * time.Millisecond):
+			vm.Interrupt <- func() {
+				panic("timed out executing Javascript")
+			}
+		case <-ctx.Done():
+			// do nothing, we finished executing
+		}
+	})
+
+	// Set the subject of the expression in a variable called $ as a simple handle to access
+	// everything.
+	_ = vm.Set("$", subject)
+
+	// Evaluate the source (eg. the script) against the subject, set above.
+	outResult, err := vm.Run(source)
+
+	if err != nil {
+		return outResult, errors.Wrap(errors.New("failed to evaluate JS against source data"), err.Error())
+	}
+
+	return outResult, nil
+
+}
+
+func EvaluateArray[ReturnType any](ctx context.Context, source string, subject any) ([]ReturnType, error) {
+	result, err := EvaluateJavascript(ctx, source, subject)
+	if err != nil {
+		// This is an expected situation: lots of Javascript code will fail depending on the
+		// input it receives.
+
+		// For now, while we do nothing with the errors (not what we'll always do) we should
+		// log only with an event that we can easily search for.
+		fmt.Println("Error when evaluating javascript", err)
```

Comment:
Can you include the source string in the error message? I think that'll make it easier to debug